File: lib/src/validators.dart
class Validators {
  // Pattern to catch points (commas and dots).
  static final points = RegExp(r"[,.]");

  // Patterns to catch nums and letters in operation.
  static final nums = RegExp(r"[0-9]"), letters = RegExp(r"[A-Za-z]");

  // Patterns to catch operation signs in expression.
  static final plusMinus = RegExp(r"[-+]"), multDivPer = RegExp(r"[/*%]");

  // Pattern to catch parentheses.
  static final parentheses = RegExp(r"[()]");

  // Checks if given char is num or not.
  static bool isNum(String c) => nums.hasMatch(c);

  // Checks if given char is plus or minus.
  static bool isPlusOrMinus(String c) => plusMinus.hasMatch(c);

  // Checks if given char is multiplication, division or percentage sign.
  static bool isNummable(String c) => !multDivPer.hasMatch(c);

  // Checks if given char is dot or comma.
  static bool isPoint(String c) => points.hasMatch(c);

  // Checks if given part is completed or not.
  static bool isNotCompletedPart(String p) {
    return multDivPer.hasMatch(p) || p.contains('+') || parentheses.hasMatch(p);
  }

  // Checks if given char is num or not.
  static bool isNumOrPoint(String c) => nums.hasMatch(c) || points.hasMatch(c);

  // Checks if given char is parentheses or not.
  static bool isPr(String c) => parentheses.hasMatch(c);

  static bool isOpeningPr(String c) => c == '(';

  static bool isClosingPr(String c) => c == ')';
}

File: pkg/validators.go
package pkg

import "regexp"

// General regular expressions for validation functions.
var (
	numRegExp = regexp.MustCompile(`\d`)
)

// IsOperationSign checks if given char is operation sign or not.
// Operation signs are - "plus", "minus", "product", "division" etc.
func IsOperationSign(ch string) bool {
	return IsPlusOrMinus(ch) || IsProdOrDiv(ch)
}

// IsNumber checks if given char is digit number or not.
func IsNumber(ch string) bool {
	return numRegExp.MatchString(ch)
}

// IsPoint checks if given char is "dot" or "comma".
func IsPoint(ch string) bool {
	return ch == "." || ch == ","
}

// IsPlusOrMinus checks if given char is "plus" sign or "minus".
func IsPlusOrMinus(ch string) bool {
	return ch == "+" || ch == "-"
}

// IsProdOrDiv checks if given char is "product" sign or "divide".
func IsProdOrDiv(ch string) bool {
	return ch == "*" || ch == "•" || ch == "/" || ch == ":"
}
